helix-term/src/config.rs

```diff
@@ -43,11 +35,68 @@ impl Display for ConfigLoadError {
 }
 
 impl Config {
+    // avoids confusion between a default rust struct and default helix keys
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> Self {
+        Config {
+            theme: None,
+            unbind_default_keys: false,
+            keys: Default::default(),
+            editor: Default::default(),
+        }
+    }
+    pub fn with_default_keys() -> Self {
+        Config {
+            theme: None,
+            unbind_default_keys: false,
+            keys: default_keys(),
+            editor: helix_view::editor::Config::default(),
+        }
+    }
+
+    /// Overwrite (or add) other's keys into our keymap
+    pub fn merge_keys(mut self, mut other: Self) -> Self {
+        for (mode, keys) in self.keys.iter_mut() {
+            if let Some(other_keys) = other.keys.remove(mode) {
+                keys.merge(other_keys);
+            }
+        }
+
+        self
+    }
+
+    /// Take other's keys and merge ours into them
+    pub fn merge_into_keys(mut self, mut other: Self) -> Self {
+        std::mem::swap(&mut self.keys, &mut other.keys);
+        self.merge_keys(other)
+    }
+
+    /// Replace our keys for a mode with other's keys for that mode.
+    /// Retains existing keys if other does not redefine them
+    pub fn replace_keys(mut self, other: Self) -> Self {
+        for (mode, keys) in other.keys.into_iter() {
+            self.keys.insert(mode, keys);
+        }
+
+        self
+    }
+
+    /// Take other's keys and replace our modes into them
+    pub fn replace_into_keys(mut self, mut other: Self) -> Self {
+        std::mem::swap(&mut self.keys, &mut other.keys);
+        self.replace_keys(other)
+    }
+
     pub fn load(config_path: PathBuf) -> Result<Config, ConfigLoadError> {
         match std::fs::read_to_string(config_path) {
-            Ok(config) => toml::from_str(&config)
-                .map(merge_keys)
-                .map_err(ConfigLoadError::BadConfig),
+            Ok(config) => match toml::from_str::<Config>(&config) {
+                Ok(config) if config.unbind_default_keys => {
+                    Ok(config.replace_into_keys(Config::with_default_keys()))
+                }
```

Comment:
This unbinds the default keys for only those modes for which there are user defined keys in the config, i.e. with this config:

```toml
unbind-default-keys = true

[keys.normal]
"t" = "goto_definition"
```

normal mode will only have the `t` keybind while all other modes will have the default bindings. This is mentioned in the docs with 

> To remove the default bindings **for all modes defined by the user**,..._

but I think it can be made clearer. A comment in the code would also be nice.

Reply:
I can do that. Initially the plan was to have it unbind all defaults of every mode but after trying it, it wasn't intuitive. I'll add some comments and a small example for the docs

Reply:
Improved comments and docs with [119c750](https://github.com/helix-editor/helix/pull/2733/commits/119c7507144517ea62c87acc469b91c0136f9708)